programs/bpf_loader/src/syscalls/cpi.rs

```diff
@@ -9,14 +13,62 @@ use {
         },
         transaction_context::BorrowedAccount,
     },
-    std::{mem, ptr, slice},
+    std::{mem, ptr},
 };
 
+fn check_account_info_pointer(
+    vm_addr: u64,
+    expected_vm_addr: u64,
+    field: &str,
+) -> Result<(), Error> {
+    if vm_addr != expected_vm_addr {
+        debug!(
+            "Invalid account info pointer `{}': {:#x} != {:#x}",
+            field, vm_addr, expected_vm_addr
+        );
+        return Err(SyscallError::InvalidPointer.into());
+    }
+    Ok(())
+}
+
+enum VmValue<'a, 'b, T> {
```

Comment:
The Translated variant will go away yes, but I'm not sure if VmValue should go away altogether. I kinda think we should go in the opposite direction and use it for `lamports` and `owner` too, since it makes it explicit that they're vm memory and not host memory.

Reply:
Just write a comment that it should turned from `enum` to `struct` once the feature gate is activated.